validphys2/src/validphys/photon/alpha.py

```diff
@@ -0,0 +1,241 @@
+            self.q = self.q[np.isfinite(self.q)]
+            self.q.sort()
+
+            self.alpha_vec = np.array([self.alpha_em(q_) for q_ in self.q])
+            self.alpha_em = self.interpolate_alphaem
+        else:
+            raise ValueError(f"Evolution mode not recognized: {self.theory['ModEv']}")
+
+    def interpolate_alphaem(self, q):
+        r"""
+        Interpolate precomputed values of alpha_em.
+
+        Parameters
+        ----------
+        q: float
+            value in which alpha_em is computed
+
+        Returns
+        -------
+        alpha_em: float
+            electromagnetic coupling
+        """
+        return np.interp(q, self.q, self.alpha_vec)
+
+    def alpha_em(self, q):
+        r"""
+        Compute the value of the running alphaem.
+
+        Parameters
+        ----------
+        q: float
+            value in which alphaem is computed
+
+        Returns
+        -------
+        alpha_em: numpy.ndarray
+            electromagnetic coupling
+        """
+        nf, nl = self.find_region(q)
+        return self.alphaem_fixed_flavor(
+            q, self.alphaem_thresh[(nf, nl)], self.thresh[(nf, nl)], nf, nl
+        )
+
+    def alphaem_fixed_flavor_trn(self, q, alphaem_ref, qref, nf, nl):
+        """
+        Compute the running alphaem for nf fixed at NLO, using truncated method.
+        In this case the RGE for alpha_em is solved decoupling it from the RGE for alpha_s
+        (so the mixed terms are removed). alpha_s will just be unused.
+
+        Parameters
+        ----------
+        q : float
+            target scale
+        alph_aem_ref : float
+            reference value of alpha_em
+        qref: float
+            reference scale
+        nf: int
+            number of flavors
+
+        Returns
+        -------
+        alpha_em at NLO : float
+            target value of a
+        """
+        if (nf, nl) == (0, 0):
+            return alphaem_ref
+        lmu = 2 * np.log(q / qref)
+        den = 1.0 + self.betas_qed[(nf, nl)][0] * alphaem_ref * lmu
+        alpha_LO = alphaem_ref / den
+        alpha_NLO = alpha_LO * (
+            1 - self.betas_qed[(nf, nl)][1] / self.betas_qed[(nf, nl)][0] * alpha_LO * np.log(den)
+        )
+        return alpha_NLO
+
+    def alphaem_fixed_flavor_exa(self, q, alphaem_ref, qref, nf, nl):
+        """
+        Compute numerically the running alphaem for nf fixed.
+
+        Parameters
+        ----------
+        q : float
+            target scale
+        alph_aem_ref : float
+            reference value of alpha_em
+        qref: float
+            reference scale
+        nf: int
+            number of flavors
+
+        Returns
+        -------
+        alpha_em: float
+            target value of a
+        """
+        u = 2 * np.log(q / qref)
+
+        # solve RGE
+        res = solve_ivp(
+            rge, (0, u), (alphaem_ref,), args=[self.betas_qed[(nf, nl)]], method="Radau", rtol=1e-6
+        )
+        return res.y[0][-1]
```

Comment:
As far as I can see this is the only place where rge is used I think.
Please add a comment before the return on why you select [0][-1] so that it is clear for future generations :P